pkg/templatelookup/regular_test.go

```diff
@@ -212,25 +213,281 @@ func TestTemplateLookup_GetRegularTemplates_WhenSwitchModuleChannel(t *testing.T
+					Name:    moduleToInstall.Name,
+					Channel: string(shared.NoneChannel),
+					Version: "1.0.0",
+					Template: &v1beta2.TrackingObject{
+						PartialMeta: v1beta2.PartialMeta{
+							Generation: 1,
+						},
 					},
-				},
-			},
-		}).Build())
-	return templateList
+				}).Build(),
+			availableModuleTemplates: (&ModuleTemplateListBuilder{}).
+				Add(moduleToInstall.Name, "regular", "1.0.0").
+				Add(moduleToInstall.Name, "fast", "1.1.0").
+				Add(moduleToInstall.Name, string(shared.NoneChannel), "1.0.0").
+				Add(moduleToInstall.Name, string(shared.NoneChannel), "1.1.0").
+				Build(),
+			wantChannel: string(shared.NoneChannel),
+			wantVersion: "1.1.0",
+		},
+		{
+			name: "When downgrade version, then result contains error",
+			kyma: builder.NewKymaBuilder().
+				WithEnabledModule(moduleToInstall).
+				WithModuleStatus(v1beta2.ModuleStatus{
+					Name:    moduleToInstall.Name,
+					Channel: string(shared.NoneChannel),
+					Version: "2.1.0",
+					Template: &v1beta2.TrackingObject{
+						PartialMeta: v1beta2.PartialMeta{
+							Generation: 1,
+						},
+					},
+				}).Build(),
+			availableModuleTemplates: (&ModuleTemplateListBuilder{}).
+				Add(moduleToInstall.Name, "regular", "1.1.0").
+				Add(moduleToInstall.Name, "fast", "2.1.0").
+				Add(moduleToInstall.Name, string(shared.NoneChannel), "1.1.0").
+				Add(moduleToInstall.Name, string(shared.NoneChannel), "2.1.0").
+				Build(),
+			wantErrContains: "as a higher version (2.1.0) of the module was previously installed",
+		},
+	}
+
+	for _, testCase := range tests {
+		t.Run(testCase.name, func(t *testing.T) {
+			lookup := templatelookup.NewTemplateLookup(NewFakeModuleTemplateReader(testCase.availableModuleTemplates),
+				provider.NewCachedDescriptorProvider())
+			got := lookup.GetRegularTemplates(context.TODO(), testCase.kyma)
+			assert.Len(t, got, 1)
+			for key, module := range got {
+				assert.Equal(t, key, moduleToInstall.Name)
+				if testCase.wantErrContains != "" {
+					assert.Contains(t, module.Err.Error(), testCase.wantErrContains)
+				} else {
+					assert.Equal(t, testCase.wantChannel, module.DesiredChannel)
+					assert.Equal(t, testCase.wantVersion, module.ModuleTemplate.Spec.Version)
+				}
+			}
+		})
+	}
+}
+
+func TestTemplateLookup_GetRegularTemplates_WhenSwitchFromChannelToVersion(t *testing.T) {
+	moduleToInstall := moduleToInstallByVersion("module1", "1.1.0")
+
+	tests := []struct {
+		name                     string
+		kyma                     *v1beta2.Kyma
+		availableModuleTemplates v1beta2.ModuleTemplateList
+		wantVersion              string
+		wantChannel              string
+		wantErrContains          string
+	}{
+		{
+			name: "When staying with the same version, then result contains no error",
+			kyma: builder.NewKymaBuilder().
+				WithEnabledModule(moduleToInstall).
+				WithModuleStatus(v1beta2.ModuleStatus{
+					Name:    moduleToInstall.Name,
+					Channel: "regular",
+					Version: "1.1.0",
```

Comment:
can you define those versions for testing `1.1.0`, `1.0.0` as a const value so that better for us to understand which field to test here?

Reply:
I thought that just version numbers serve as a better "name" than the name itself - they're usually shorter and visually distinct. But If you think conts like "oldVersion" "newVersion" are better, I'll do that.